File: app/src/main/java/cn/winxo/toolbox/data/dao/BaseDao.java
package cn.winxo.toolbox.data.dao;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Update;
import java.util.List;

@Dao abstract class BaseDao<T> {

  @Insert(onConflict = OnConflictStrategy.REPLACE) public abstract long replace(T t);

  @Insert(onConflict = OnConflictStrategy.REPLACE) public abstract void insert(List<T> t);

  @Insert public abstract long insert(T t);

  @Update public abstract void update(T t);

  @Update public abstract void update(List<T> t);

  @Delete public abstract void deleteBatch(List<T> list);

  @Delete public abstract void delete(T t);
}

File: app/src/main/java/cn/winxo/toolbox/data/dao/BaseDao.kt
package cn.winxo.toolbox.data.dao

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Update

@Dao
abstract class BaseDao<T> {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun replace(t: T): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insert(t: List<T>)

    @Insert
    abstract fun insert(t: T): Long

    @Update
    abstract fun update(t: T)

    @Update
    abstract fun update(t: List<T>)

    @Delete
    abstract fun deleteBatch(list: List<T>)

    @Delete
    abstract fun delete(t: T)
}